// generators/blog-tag-pages.js
var blogTagPages, mustache, getPath, mostRecent, sortByDate, getSlug;


mustache = require( 'mustache' );
mostRecent = require( 'snippets/mostRecent' );
sortByDate = require( 'utils/sortByDate' );
getSlug = require( 'utils/getSlug' );


blogTagPages = function ( data ) {
	
	'use strict';

	var pages, i, j, id, post, posts, postTags, tag, tagsBySlug, mustacheData, postsByTag, path, slug, slugs;

	tagsBySlug = {};
	slugs = [];
	postsByTag = {};

	posts = data.blogposts;

	// organise posts by tag
	i = posts.length;
	while ( i-- ) {
		post = posts[i];

		postTags = post.tags || [];

		j = postTags.length;
		while ( j-- ) {
			tag = postTags[j];

			if ( !postsByTag[ tag.slug ] ) {
				postsByTag[ tag.slug ] = [];
			}

			if ( !tagsBySlug[ tag.slug ] ) {
				tagsBySlug[ tag.slug ] = tag;
			}

			postsByTag[ tag.slug ].push( post );
		}
	}

	// generate a page for each tag
	pages = [];

	for ( slug in postsByTag ) {
		path = 'blog/tags/' + slug + '.html';
		tag = tagsBySlug[ slug ];

		posts = postsByTag[ slug ].sort( sortByDate );
		
		mustacheData = {
			tag: tag,
			tagPath: path,
			pageTitle: tag.name + ' posts',
			description: 'Science: Gamed blog posts tagged with "' + tag.name + '"',
			posts: postsByTag[ slug ].sort( sortByDate )
		};


		pages[ pages.length ] = {
			path: path,
			data: mustache.render( data.templates.indexpage, mustacheData, data.templates.partials )
		};
	}


	return pages;
};


module.exports = blogTagPages;
